link/kprobe_test.go

```diff
@@ -0,0 +1,191 @@
+	testutils.SkipOnOldKernel(t, "5.0", "lifted version check for kprobes")
+
+	c := qt.New(t)
+
+	prog, err := ebpf.NewProgram(&kprobeSpec)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer prog.Close()
+
+	tp, err := Kretprobe("printk")
+	c.Assert(err, qt.IsNil)
+	c.Assert(tp.ret, qt.Equals, true) // kretprobe
+	defer tp.Close()
+
+	l, err := tp.Attach(prog)
+	c.Assert(err, qt.IsNil)
+	defer l.Close()
+}
+
+func TestKprobeTraceEvent(t *testing.T) {
+	tp, err := kprobe("printk", false)
+	if err != nil {
+		t.Fatal("creating kprobe:", err)
+	}
+	defer tp.Close()
+}
+
+// Test k(ret)probe creation using perf_kprobe PMU.
+func TestKprobeCreatePMU(t *testing.T) {
+
+	// Requires at least 4.15 (e12f03d7031a "perf/core: Implement the 'perf_kprobe' PMU")
+	testutils.SkipOnOldKernel(t, "4.15", "perf_kprobe PMU")
+
+	c := qt.New(t)
+
+	// kprobe happy path. printk is always present.
+	pk, err := pmuKprobe("printk", false)
+	c.Assert(err, qt.IsNil)
+	defer pk.Close()
+
+	c.Assert(pk.progType, qt.Equals, ebpf.Kprobe)
+	c.Assert(pk.finalizer, qt.IsNil)
+
+	pe, err := pk.perfEventOpenPMU()
+	c.Assert(err, qt.IsNil)
+	defer unix.Close(pe)
+
+	// kretprobe happy path.
+	pr, err := pmuKprobe("printk", true)
+	c.Assert(err, qt.IsNil)
+	defer pr.Close()
+
+	c.Assert(pr.progType, qt.Equals, ebpf.Kprobe)
+	c.Assert(pr.finalizer, qt.IsNil)
+
+	pe, err = pr.perfEventOpenPMU()
+	c.Assert(err, qt.IsNil)
+	defer unix.Close(pe)
+}
+
+// Test k(ret)probe creation handling pre-existing trace events.
+func TestKprobeTraceFS(t *testing.T) {
+	c := qt.New(t)
+
+	symbol := "printk"
+	group := mustRandomGroup("ebpftest")
+
+	// First call using this group and symbol will write to <tracefs>/kprobe_events.
+	// The TraceEvent must contain a finalizer to remove the record on Close.
+	te, err := tracefsKprobe(group, symbol, false)
+	c.Assert(err, qt.IsNil)
+	defer te.Close()
+	c.Assert(te.finalizer, qt.Not(qt.IsNil))
+	c.Assert(te.progType, qt.Equals, ebpf.Kprobe)
+	c.Assert(te.tracefsID, qt.Not(qt.Equals), 0)
+
+	// Second invocation. Identical trace event already exists,
+	// so must not contain a finalizer
+	te, err = tracefsKprobe(group, symbol, false)
+	c.Assert(err, qt.IsNil)
+	defer te.Close()
+	c.Assert(te.finalizer, qt.IsNil)
+}
+
+// Test k(ret)probe creation unconditionally writing to <tracefs>/kprobe_events.
+func TestKprobeCreateTraceFS(t *testing.T) {
+	c := qt.New(t)
+
+	symbol := "printk"
+	group := mustRandomGroup("ebpftest")
+
+	// Create a kprobe.
+	err := createTraceFSKprobeEvent(group, symbol, false)
+	c.Assert(err, qt.IsNil)
+	// Create a kretprobe on the same symbol using the same group.
+	err = createTraceFSKprobeEvent(group, symbol, true)
+	c.Assert(err, qt.IsNil)
+
+	// Attempt to create an identical kprobe using tracefs,
+	// expect it to fail with os.ErrExist.
+	err = createTraceFSKprobeEvent(group, symbol, false)
+	c.Assert(err, qt.Not(qt.IsNil))
+	// Don't bail out here, the close(s) below needs to fire to avoid dangling
```

Comment:
Ah, didn't occur to me that it would be possible, because we're reusing `err` here (which bites when deferring closures). I wanted to make sure those calls returned `nil`. Since it's a test suite, we can easily do `defer c.Assert(closeTraceFSKprobeEvent(group, symbol, false), qt.IsNil)` instead, which won't gobble up the errors. :100: